indexer/services/vulcan/src/handlers/order-replace-handler.ts

```diff
@@ -0,0 +1,393 @@
+import { logger, runFuncWithTimingStat, stats } from '@dydxprotocol-indexer/base';
+import { createSubaccountWebsocketMessage, KafkaTopics } from '@dydxprotocol-indexer/kafka';
+import {
+  OrderFromDatabase,
+  OrderTable,
+  PerpetualMarketFromDatabase,
+  blockHeightRefresher,
+  perpetualMarketRefresher,
+  protocolTranslations,
+} from '@dydxprotocol-indexer/postgres';
+import {
+  CanceledOrdersCache,
+  OrderbookLevelsCache,
+  placeOrder,
+  PlaceOrderResult,
+  StatefulOrderUpdatesCache,
+  convertToRedisOrder,
+  removeOrder,
+  RemoveOrderResult,
+} from '@dydxprotocol-indexer/redis';
+import {
+  getOrderIdHash,
+  isLongTermOrder,
+  isStatefulOrder,
+  ORDER_FLAG_SHORT_TERM,
+  requiresImmediateExecution,
+} from '@dydxprotocol-indexer/v4-proto-parser';
+import {
+  IndexerOrder,
+  IndexerOrderId,
+  OffChainUpdateV1,
+  OrderPlaceV1_OrderPlacementStatus,
+  OrderReplaceV1,
+  OrderUpdateV1,
+  RedisOrder,
+} from '@dydxprotocol-indexer/v4-protos';
+import Big from 'big.js';
+import { IHeaders, Message } from 'kafkajs';
+
+import config from '../config';
+import { redisClient } from '../helpers/redis/redis-controller';
+import { sendMessageWrapper } from '../lib/send-message-helper';
+import { Handler } from './handler';
+
+/**
+ * Handler for OrderReplace messages.
+ * The behavior is as follows:
+ * - Remove the old order
+ *  - this is done using the `removeOrder` function from the `redis` package
+ * - Add the new order to the OrdersCache, OrdersDataCache, and SubaccountOrderIdsCache
+ *  - this is done using the `placeOrder` function from the `redis` package
+ *  - Remove the order from the CanceledOrdersCache if it exists
+ * - Because the order is a stateful order, attempt to remove any cached order update from the
```

Comment:
Is it assumed that the order is a stateful order? If so, the function logic can be simplified a lot and a lot of conditionals related to short term orders can be removed.

Reply:
I think it'd be good to have memclob switch to sending this message on short term order replacements instead of the current flow. But currently only stateful vault orders would go through this

Reply:
Nit on the comment, then, it should be `If the order is a stateful order...`

Reply:
Changed!